javascript/401-code-challenges/hash-tables/hash-tables.test.js

```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+const { Hashtable } = require('./index');
+
+
+describe('Hash tables', () => {
+
+  test('Setting a key/value to the hashtable, resulting in the value being in the data structure', () => {
+    const hashtable = new Hashtable();
+    hashtable.set('key1', 'value1');
+
+    expect(hashtable.get('key1')).toEqual('value1');
+  });
+
+
+  test('Retrieving based on a key returns the value stored', () => {
+    const hashtable = new Hashtable();
+    hashtable.set('key2', 'value2');
+
+    expect(hashtable.get('key2')).toEqual('value2');
+  });
+
+
+  test('Successfully returns null for a key that does not exist in the hashtable', () => {
+    const hashtable = new Hashtable();
+
+    expect(hashtable.get('nonexistent')).toBeUndefined();
+  });
+
+
+  test('Successfully returns a list of all unique keys that exist in the hashtable', () => {
+    const hashtable = new Hashtable();
+    hashtable.set('key1', 'value1');
+    hashtable.set('key2', 'value2');
+    hashtable.set('key3', 'value3');
+    hashtable.set('key4', 'value4');
+    const expectedKeys = ['key1', 'key2', 'key3', 'key4'];
+
+    expect(hashtable.keys().sort()).toEqual(expect.arrayContaining(expectedKeys));
+  });
+
+
+  test('Successfully handle a collision within the hashtable', () => {
+    const hashtable = new Hashtable();
+    hashtable.set('abc', 'value5');
+    hashtable.set('cba', 'value6');
+
+    expect(hashtable.get('abc')).toEqual('value5');
+    expect(hashtable.get('cba')).toEqual('value6');
+  });
+
+
+  test('Successfully retrieve a value from a bucket within the hashtable that has a collision', () => {
+    const hashtable = new Hashtable();
+    hashtable.set('xyz', 'value7');
+    hashtable.set('zyx', 'value8');
+
+    expect(hashtable.get('xyz')).toEqual('value7');
+    expect(hashtable.get('zyx')).toEqual('value8');
+  });
```

Comment:
...while these pairs of keys had the same hash with the algorithm we used in class, they are not the same with your algorithm (unless you created a hashtable with size 2<sup>n</sup> where n < 8, but that's a discussion for another time). You might just make a hashtable with size 1 so you are guaranteed collisions. At any rate, your code does deal with collisions properly, so it's just making sure the tests are actually testing for it.